vault.go

```diff
@@ -526,16 +520,19 @@ func newVaultClient(opts *config) (*api.Client, error) {
 		renewPeriod := tokenttl / 2
 		go func() {
 			for {
-				if renewPeriod < 1*time.Second {
-					glog.Fatalf("fatal: token renew period is <1s, aborting")
-				}
 				glog.Infof("scheduling token renew in %v", renewPeriod)
 				<-time.After(renewPeriod)
 
 				glog.Infof("attempting token renew")
 				newtokeninfo, err := client.Auth().Token().RenewSelf(0)
 				if err != nil {
 					renewPeriod = renewPeriod / 2
+
+					// Keep renewPeriod >= 30 seconds (plus a variable jitter time in the 0-15s range)
+					if renewPeriod < 30*time.Second {
+						renewPeriod = 30*time.Second + getDurationWithin(0, 15)
```

Comment:
Makes sense, thanks for the great comment. 

We're making a trade off here between Vault load and the risk of our token expiring. The default `vault-sidekick `code chooses to reduce the risk of our token expiring, at the expense of increased Vault load - i.e we always set the renew period to something _before_ the token expires (`renewPeriod = renewPeriod / 2`). This change we're making here chooses the opposite: we decide to set the renew period to a time to something _after_ the token expires. In doing this, we reduce the load on Vault, but we increase the risk of our token expiring.

I think this completely makes sense, given we have a whole week before an expired Vault token will be become a problem (e.g. for cert expiry) so it makes sense to choose reduced Vault load over token expiry.  I think it'd be a good idea to make decision explicit in the code. What do you think of adding a comment in to explicitly show this here?

Reply:
> This change we're making here chooses the opposite: we decide to set the renew period to a time to something after the token expires. In doing this, we reduce the load on Vault, but we increase the risk of our token expiring.

I'm not sure I'm following as we are still setting the `renewPeriod` time to `tokenttl/2` (and it still gets halved when the renew operation fails) so:
* in normal conditions the token will get renewed before it expires
* in case of repeated failures the `renewPeriod` duration will eventually reach a value below 30 seconds (and so it will get assigned one between 30-45s)

This means that the case where the renew period is set to something after the token expires will only happen after vault-sidekick has failed multiple times to refresh the token

Reply:
Yeah, that's right 👍 

I'm specifically talking about this case:
> in case of repeated failures the renewPeriod duration will eventually reach a value below 30 seconds (and so it will get assigned one between 30-45s)

I.e. when we repeatedly fail to renew tokens, at some point we'll begin scheduling token renewals for _after_ the token expires. This says "we care about renewing the token, but we care more about making sure we don't overload vault"

Reply:
Got it, I'll update the comment 👍